staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -17,50 +17,96 @@ limitations under the License.
 package meta
 
 import (
+	"errors"
 	"fmt"
 	"reflect"
+	"sync"
 
 	"k8s.io/apimachinery/pkg/conversion"
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
-// IsListType returns true if the provided Object has a slice called Items
+var (
+	// isListCache maintains a cache of types that are checked for lists
+	// which is used by IsListType.
+	// TODO: remove and replace with an interface check
+	isListCache = struct {
+		lock   sync.RWMutex
+		byType map[reflect.Type]bool
+	}{
+		byType: make(map[reflect.Type]bool, 1024),
+	}
+)
+
+// IsListType returns true if the provided Object has a slice called Items.
+// TODO: Replace the code in this check with an interface comparison by
+//   creating and enforcing that lists implement a list accessor.
 func IsListType(obj runtime.Object) bool {
-	// if we're a runtime.Unstructured, check whether this is a list.
-	// TODO: refactor GetItemsPtr to use an interface that returns []runtime.Object
-	if unstructured, ok := obj.(runtime.Unstructured); ok {
-		return unstructured.IsList()
+	switch t := obj.(type) {
+	case runtime.Unstructured:
+		return t.IsList()
 	}
+	t := reflect.TypeOf(obj)
 
-	_, err := GetItemsPtr(obj)
-	return err == nil
+	isListCache.lock.RLock()
+	ok, exists := isListCache.byType[t]
+	isListCache.lock.RUnlock()
+
+	if !exists {
+		_, err := getItemsPtr(obj)
+		ok = err == nil
+
+		// cache only the first 1024 types
+		isListCache.lock.Lock()
+		if len(isListCache.byType) < 1024 {
```

Comment:
It seems before acquiring the lock, we can check the length.
Since there is no eviction, the length would only grow.
If the length reaches 1024, there is no need to acquire the lock.

Reply:
Created #75630

Reply:
len() is not thread safe. I don't think you can do it.

Reply:
you can do it inside the RLock above

Reply:
Yes, using a RWLock first use RLock() for the len check (read) and upgrade to write lock via Lock() only the map update.

Reply:
Does go support RWLock upgrade (from read to write) ?

If not, using the read lock doesn't seem necessary.
Since the length of map only increases, if the map length reaches 1024 outside the lock, it wouldn't be lower than 1024 inside the lock.

Reply:
I don't think we can assume that a len check on a map being unsafely concurrently updated will return a sane value, even if you're only ever growing the map size.

In go RWLock isn't directly upgradable, so we'd just RLock around the list size read, then release the RLock, and then separately Lock around the write.

Reply:
My comment "len() is not thread safe. I don't think you can do it." was not about the code, it was about Ted's comment. Yes, if there is a RLock you can do it, but as Joe described it clearly, you cannot check the len outside of a lock hence #75630 does not work.

Reply:
That was closed.

Please take a look at #75702